### ISSUE TITLE
[QUESTION] Will helm wait pass through before resources status not changed yet?

### ISSUE THREAD

**Issue by dayeguilaiye** [helm/helm#12847]
I have a CRD and it's operator, when CR updated, operator will modify the status of this CR to `updating` from `done`. After all operation done, operator will change the status into `done` again.

I wrote another go client program to modify this CR and wait until it done. But when I query the status of this CR just after send the update request to k8s, the status is still `done`(the operator is not fast enough to change it to `updating` before this), so the program will consider it already updated, but in fact, it wasn't.

Now, I will check the time of the condition in status. This can fix the problem.

But I wondering how helm deal with this problem, I found the code:
https://github.com/helm/helm/blob/15d116b8548f59959fe9f2fafdaf7c6de8db3ea0/pkg/kube/ready.go#L230-L238

It didn't check the time of condition, but seems the problem didn't happened to helm. Anyone can tell me the reason, please?

Thanks a lot!

**Comment by gjenkins8** [comment]
Helm, in general, only knows how to wait for a certain built-in objects. It won't wait for Custom Resources, as it doesn't know what "ready" means for those.

Helm would do well to add a more generic wait mechanism (e.g. jsonpath conditions). But has not had resources do attempt to do so.

I think that answers the question here?

**Comment by dayeguilaiye** [comment]
@gjenkins8 Thanks for your reply!

I understand why helm don't wait for custom resources "done". My question is why the logic in helm when waiting built-in objects don't have problem like what happened to me.

For example: There's a sucess pod deployed by helm, then i use helm to upgrade. After helm send update request to k8s, it will query the status of this pod immediately. What if the controller of pod is not soon enough to change the status(condition) of this pod yet?
Should helm think the status is already ready, then break the waiting?

But as I know, this problem didn't happened to helm. So I wander if there's any different between custom resources and built-in resources? For example, is status (condition) changed before the update request returned when the resource is built-in resource(which is different from CR)?

**Comment by gjenkins8** [comment]
> For example: There's a "ready" pod deployed by helm, then I use helm to upgrade it. After helm send update pod request to k8s, helm will query the status of this pod immediately. What if the controller of pod is not soon enough to change the status(condition) of this pod yet?
Should helm think the status is already ready, then break the waiting?

Yes. Once Helm detects the pod, or deployment, etc, is ready. It will stop waiting. In general, Helm must expect it is the only operator during its operations. It has not ability to synchronize with external processes.

**Comment by dayeguilaiye** [comment]
I thought helm don't have those problem because there are some different between k8s built-in resources and custom resources.
But I found this problem happens to helm even upgrading built-in resource like pod too.

I started a new issue :https://github.com/helm/helm/issues/12861
